group/assignment-6/src/cars/RearrangingCars.java

```diff
@@ -0,0 +1,323 @@
+            if (!endState.getSpotOf(car).equals(startState.getSpotOf(car))) {
+                carsToBeMoved.add(car);
+            }
+        }
+
+        return carsToBeMoved;
+    }
+
+    private static int getNumOfCarsInEndLocation(
+            Map<Car, ParkingSpot> configuration, ParkingLot endState) {
+        int numOfCarsInEndLocation = 0;
+
+        for (Car car : configuration.keySet()) {
+            if ((car != null) && configuration.get(car).equals(endState.getSpotOf(car))) {
+                numOfCarsInEndLocation++;
+            }
+        }
+
+        return numOfCarsInEndLocation;
+    }
+
+    private static void updateStartStateConfiguration(
+            Map<Car, ParkingSpot> configuration, ParkingLot startState) {
+        for (ParkingSpot spot : startState.getSpots()) {
+            configuration.put(startState.getCarAt(spot), spot);
+        }
+    }
+
+    private static void validateParkingLots(ParkingLot startState, ParkingLot endState)
+            throws IllegalArgumentException {
+        if ((startState == null) || (endState == null)) {
+            throw new IllegalArgumentException("You must enter defined parking lots.");
+        }
+
+        if (!startState.equals(endState)) {
+            throw new IllegalArgumentException(
+                    "The parking lots must contain the same parking spots and cars.");
+        }
+    }
+
+    // to be potentially modified based on allowedParking constraints format (to be decided)
+    private static void validateParkingLots(
+            ParkingLot startState, ParkingLot endState, Map<ParkingSpot, List<Car>> allowedParking)
+            throws RuntimeException {
+        validateParkingLots(startState, endState);
+        validateAllowedParkingConstraints(startState, allowedParking);
+    }
+
+    // to be potentially modified based on allowedParking constraints format (to be decided)
+    private static void validateAllowedParkingConstraints(
+            ParkingLot parkingLot, Map<ParkingSpot, List<Car>> allowedParking)
+            throws IllegalArgumentException {
+        Collection<Car> cars = parkingLot.getCars();
+        Collection<ParkingSpot> spots = parkingLot.getSpots();
+
+        for (Entry<ParkingSpot, List<Car>> currentConstraint : allowedParking.entrySet()) {
+            if (!spots.contains(currentConstraint.getKey())) {
+                throw new IllegalArgumentException(
+                        "Each parking spot in the allowed parking constraints has to be contained in the given parking lot.");
+            }
+
+            if (!cars.containsAll((currentConstraint.getValue()))) {
+                throw new IllegalArgumentException(
+                        "All mapped cars in the allowed parking constraints have to be contained in the given parking lot.");
+            }
+        }
+    }
+
+
+    private static void getAllPossibleSequencesUtil(ParkingLotConfigurations trie, Map<ParkingSpot, Car> currentState,
```

Comment:
```suggestion
    private static void getAllPossibleSequencesUtil(ParkingLotConfigurations<ParkingSpot> trie, Map<ParkingSpot, Car> currentState,
```
It is parameterized with `ParkingSpot` class